Make exist-xqts runner compatible with 7.0.0-SNAPSHOT of exist-db

line-o opened the pull request:
This is a breaking change as older versions of exist-db cannot use this xqts-runner anymore 

- load exist-core snapshots from Github Maven package repository
- update dependencies
- fix codebase to be compatible with current develop HEAD
- fix module import
- address some linter warnings
- due to Licensing issues the dependency on Akka was replaced with [Apache Pekko](https://pekko.apache.org)

adamretter committed:
Update for eXist-db 7.0.0-SNAPSHOT

adamretter committed:
Update dependency

adamretter committed:
Simplify logging

adamretter committed:
chore: Bump Scala, Sbt, and plugin versions

BREAKING CHANGE: build target is now Java21

line-o commented:
for compilation to work again we need to point the build process to a more recent version of exist v7.0.0-SNAPSHOT

adamretter commented:
> older versions of exist-db cannot use this xqts-runner anymore

@line-o There are different versions of xqts-runner that target 6.x.x and various 7.x.x - you just need to choose the right version. You can find the version of eXist-db supported by looking in the build.sbt of the various versions of xqts-runner

line-o commented:
@adamretter I tried all published versions 1.3.0-SNAPSHOT, 1.3.0 and 1.4.0 none of them worked. This branch built against a current develop head does.

adamretter commented:
@line-o I am afraid you looked in the wrong place. If you want 7.0.0-SNAPSHOT compatibility it has been here for years - https://github.com/eXist-db/exist-xqts-runner/tree/compat/eXist-7.0.0-SNAPSHOT

line-o commented:
Now that this finally passed the smoke test again a little cleanup is needed before it can be marked as ready.

line-o committed:
ci: update github action workflow

line-o committed:
fix: use correct Type and module import

line-o committed:
chore: formatting, linter suggestions and typos

line-o committed:
chore: update conf.xml to current version

line-o committed:
chore(deps): update library dependencies

* mark exist-core dependency as "changing" to enforce SNAPSHOT updates
* update slf4j and log4j-slf4j2-impl to the same version that exist uses at the moment
* update akka-actor to 2.8.1-M1 to fix a vulnerability
* update all other dependencies to the latest non-breaking update

line-o committed:
build: add assemblyMergeStragegies to fix build, add linter

line-o committed:
build: switch to github maven repository for snapshots

Dependencies like exist-core are published to the github maven repository.
It does _require_ authentication. For this to work a Github personal access token needs to be set.
This can be done in either `~/.ivy2/.credentials` or setting the environment variable "GITHUB_TOKEN" (used in the github action).

line-o committed:
docs(readme): update prerequisites and publishing

line-o committed:
chore: add Juri Leino as a developer

line-o committed:
chore: ignore IDE and OS specific meta files

line-o committed:
chore: set version to 2.0.0-SNAPSHOT

adamretter commented on .github/workflows/ci.yml:
```diff
@@ -9,28 +9,25 @@ jobs:
       matrix:
         os: [ubuntu-latest, macos-latest, windows-latest]
-        jdk: [8, 11]
+        jdk: [21]
```
JDK version is incorrect.

line-o replied:
exist-7.0.0-SNAPSHOT has java 21 as its baseline

adamretter commented on .github/workflows/ci.yml:
```diff
@@ -9,28 +9,25 @@ jobs:
       - name: Test
         shell: bash
         run: target/scala-2.13/exist-xqts-runner-assembly-*-SNAPSHOT.jar --xqts-version HEAD --test-set fn-current-date
-      - name: Cleanup before cache
```
This stuff should not be cached,

line-o replied:
The moment that caches become a problem I will make sure to have a look. At the moment the project seems to be build fine. Also, please note that the entire setup and caching strategy has changed.

adamretter commented on README.md:
````diff
@@ -12,17 +12,22 @@ This application executes a W3C XQTS against an embedded eXist-db server.
 
 ```scala
-val existV = "5.3.0"
+val existV = "7.0.0-SNAPSHOT"
````
This won't work in production as snapshots are not permitted

line-o replied:
I do not understand why this cannot be written in the readme. Should there be an additional sentence? To which production are you referring to?

adamretter commented on build.sbt:
```diff
@@ -36,30 +36,35 @@ developers := List(
-    "com.github.scopt" %% "scopt" % "4.0.1",
-    "org.typelevel" %% "cats-effect" % "3.4.5",
-    //"com.fasterxml" %	"aalto-xml" % "1.1.0-SNAPSHOT",
+    "com.typesafe.akka" %% "akka-actor" % "2.8.1-M1",
```
Akka should not be updated here due to the licensing problem.

line-o replied:
Thanks for bringing this up @adamretter 
For clarification: the Akka project changed to a source available license (!) and 2.8 cannot be used freely
I will therefore migrate this project to use the open source fork Apache Pekko

adamretter replied:
@line-o There is absolutely no value in that, or really even in maintaining this project at all. Better to stay with the last free Akka version.

adamretter commented on build.sbt:
```diff
@@ -36,30 +36,35 @@ developers := List(
     "net.sf.saxon" % "Saxon-HE" % "9.9.1-8",
-    "org.exist-db" % "exist-core" % existV,
-    "org.xmlunit" % "xmlunit-core" % "2.9.1",
+    "org.exist-db" % "exist-core" % existV changing() exclude("org.eclipse.jetty.toolchain", "jetty-jakarta-servlet-api"),
```
Why are these excludes needed?

line-o replied:
Jakarta APIs will otherwise be in the project in conflicting versions

adamretter replied:
It should not conflict. Why does that conflict?

line-o replied:
I think it is also important to mention that this exclude was added in this commit f13591bc882cf8f2ca64c90b628fc15adc8f6e3a

adamretter commented on build.sbt:
```diff
@@ -73,12 +78,13 @@ excludeDependencies ++= Seq(
-  "eXist-db Releases" at "https://repo.evolvedbinary.com/repository/exist-db/",
-  "eXist-db Snapshots" at "https://repo.evolvedbinary.com/repository/exist-db-snapshots/",
-  "eXist-db Maven Repo" at "https://raw.github.com/eXist-db/mvn-repo/master/"
+  "eXist-db Releases" at "https://repo.exist-db.org/repository/exist-db/",
```
This should not be here. eXist-db releases have been published to the authoritative Maven Central for many years now.

line-o replied:
Maven central was never listed. But it does make sense to add it as well.

line-o replied:
According to https://www.scala-sbt.org/1.x/docs/Resolvers.html Maven Central is added by default so we do not need to add it here explicitly

adamretter commented on src/main/scala/org/exist/xqts/runner/Logger.scala:
```diff
@@ -0,0 +1,122 @@
+
+import org.slf4j.{Logger => SLF4JLogger, LoggerFactory => SLF4JLoggerFactory}
+
+final class Logger(val logger: SLF4JLogger) {
```
Why is this even needed?

line-o replied:
This class was introduced by you in eeb7a9de086798ddad828e72fb4b28acc5593c9b

adamretter reviewed (commented):
This targets the wrong branch, and is redundant as the work was already done years ago and can be found in the branch https://github.com/eXist-db/exist-xqts-runner/tree/compat/eXist-7.0.0-SNAPSHOT. In addition this project is redundant entirely as it has been replaced by XTH.

line-o commented:
@adamretter out of curiosity, what is XTH?

adamretter commented:
> out of curiosity, what is XTH?

@line-o XTH = X Test Harness. It replaced this some time ago.

line-o commented:
Do you have a link perhaps? @adamretter I cannot find any information on that.

line-o committed:
chore: move from Akka to Pekko

line-o commented:
> This targets the wrong branch, and is redundant as the work was already done years ago and can be found in the branch [...]

This PR is built on top of the other branch and supersedes it. It targets `main` as this will be the next major version, if accepted.

adamretter commented:
> This PR is built on top of the other branch and supersedes it. It targets main as this will be the next major version, if accepted.

That is not how this project is setup or structured.

line-o commented:
> That is not how this project is setup or structured.

I am not sure what you mean by this. If you want to suggest a way forward please do so, otherwise I would like to continue with the current plan.

dizzzz commented:
PR looks solid to me; as stated we can correctly run the suite now, that is what counts in the end.
Probably there is an (yet unknown) alternative as stated , but this is what we can use right now.

dizzzz reviewed (commented):
LGTM

adamretter commented:
> I am not sure what you mean by this. If you want to suggest a way forward please do so, otherwise I would like to continue with the current plan.

The project was setup to have two branches - one for eXist-db and Elemental 6.x.x, and the other one for 7.x.x

line-o commented:
It will still be that way. Just a slightly different approach:

- main: contains the version compatible with latest develop HEAD 

- another branch - its name is not determined yet - will be used to publish legacy releases, if necessary 

Please Note: Because tests are run against a version of existdb we only guarantee compatibility with this processor

adamretter commented:
@line-o Understood, but switching the main branch doesn't make sense until there is a release of eXist-db 7.0.0 as otherwise it cannot be used in production.

line-o commented:
For me main is the branch I expect to get the current latest state from.

adamretter commented:
> For me main is the branch I expect to get the current latest state from.

For me, what comes from `main` should be usable.
